drivers/gpio/gpio_pca95xx.c

```diff
@@ -440,12 +456,150 @@ static int gpio_pca95xx_port_toggle_bits(struct device *dev, uint32_t mask)
 	return ret;
 }
 
+#ifdef CONFIG_GPIO_PCA95XX_INTERRUPT
+static void gpio_pca95xx_work_handler(struct k_work *work)
+{
+	struct gpio_pca95xx_drv_data *ddata =
+			CONTAINER_OF(work, struct gpio_pca95xx_drv_data, work);
+	uint16_t status;
+	int i = 0;
+
+	if (read_port_regs(ddata->exp_dev, REG_INT_STATUS_PORT0,
+			   &status) != 0) {
+		return;
+	}
+
+	while (status) {
+		if (status & 0x1) {
+			gpio_fire_callbacks(&ddata->cb_list,
+					    ddata->exp_dev,
+					    BIT(i));
+		}
+		status >>= 1;
+		i++;
+	}
+}
+
+static void gpio_pca95xx_global_callback(struct device *port,
+					 struct gpio_callback *cb, uint32_t pin)
+{
+	struct gpio_pca95xx_drv_data *ddata =
+			CONTAINER_OF(cb, struct gpio_pca95xx_drv_data, gpio_cb);
+
+	ARG_UNUSED(port);
+	ARG_UNUSED(pin);
+
+	k_work_submit(&ddata->work);
+}
+
+static int gpio_pca95xx_interrupt_init(struct device *dev)
+{
+	struct gpio_pca95xx_drv_data *ddata = dev->data;
+	const struct gpio_pca95xx_config *cfg = dev->config;
+	int ret;
+
+	ddata->gpio = device_get_binding(cfg->gpio_port);
+	if (!ddata->gpio) {
+		LOG_ERR("Gpio controller %s not found.", cfg->gpio_port);
+		return -EINVAL;
+	}
+
+	ddata->work.handler = gpio_pca95xx_work_handler;
+
+	ret = gpio_pin_configure(ddata->gpio, cfg->int_pin,
+				 GPIO_INPUT | cfg->int_flags);
+	if (ret) {
+		LOG_ERR("gpio %d input config failed err %d.",
+			 cfg->int_pin, ret);
+		return ret;
+	}
+
+	gpio_init_callback(&ddata->gpio_cb, gpio_pca95xx_global_callback,
+			   BIT(cfg->int_pin));
+
+	ret = gpio_add_callback(ddata->gpio, &ddata->gpio_cb);
+	if (ret) {
+		LOG_ERR("gpio add callback err %d.", ret);
+		return ret;
+	}
+
+	ret = gpio_pin_interrupt_configure(ddata->gpio, cfg->int_pin,
+					   GPIO_INT_EDGE_FALLING);
+
+	return ret;
+}
+
+static int gpio_pca95xxx_manage_callback(struct device *dev,
+					 struct gpio_callback *cb,
+					 bool set)
+{
+	struct gpio_pca95xx_drv_data *ddata = dev->data;
+
+	return gpio_manage_callback(&ddata->cb_list, cb, set);
+}
+
+static uint32_t gpio_pca95xxx_get_pending_int(struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	return -ENOTSUP;
+}
+#endif
+
 static int gpio_pca95xx_pin_interrupt_configure(struct device *dev,
-						  gpio_pin_t pin,
-						  enum gpio_int_mode mode,
-						  enum gpio_int_trig trig)
+			gpio_pin_t pin, enum gpio_int_mode mode,
+			enum gpio_int_trig trig)
 {
+#ifdef CONFIG_GPIO_PCA95XX_INTERRUPT
```

Comment:
This isn't right, with the other conditional branch always returning `-ENOTSUP`, because it's OK to return 0 if interrupt support is configured but the configuration is to not use interrupts.  Look at `gpio_sx1509b.c` and do something like:

        int rc = 0;

        if (!IS_ENABLED(CONFIG_GPIO_PCA95XX_INTERRUPT)
            && (mode != GPIO_INT_MODE_DISABLED)) {
                return -ENOTSUP;
        }
    #ifdef CONFIG_GPIO_PCA95XX_INTERRUPT
    ...
    #endif 
    return 0